c_emulator/riscv_sim.cpp

```diff
@@ -40,62 +41,83 @@ enum {
   OPT_PRINT_ISA,
 };
 
-static bool do_show_times = false;
+bool do_show_times = false;
+bool do_print_version = false;
+bool do_print_build_info = false;
+bool do_print_default_config = false;
 bool do_print_dts = false;
 bool do_validate_config = false;
 bool do_print_isa = false;
 
-char *config_file = NULL;
-char *term_log = NULL;
-static const char *trace_log_path = NULL;
+std::string config_file;
+std::string term_log;
+std::string trace_log_path;
 FILE *trace_log = NULL;
-char *dtb_file = NULL;
+std::string dtb_file;
 unsigned char *dtb = NULL;
 size_t dtb_len = 0;
+int rvfi_dii_port = 0;
 std::optional<rvfi_handler> rvfi;
+std::vector<std::string> elfs;
 
-char *sig_file = NULL;
+std::string sig_file;
 uint64_t mem_sig_start = 0;
 uint64_t mem_sig_end = 0;
-int signature_granularity = 4;
+const int DEFAULT_SIGNATURE_GRANULARITY = 4;
+int signature_granularity = DEFAULT_SIGNATURE_GRANULARITY;
 
 bool config_print_instr = true;
 bool config_print_reg = true;
 bool config_print_mem_access = true;
 bool config_print_platform = true;
-bool config_use_abi_names = false;
 bool config_print_rvfi = false;
 bool config_print_step = false;
+
+bool config_use_abi_names = false;
 bool config_enable_rvfi = false;
 
-void set_config_print(char *var, bool val)
-{
-  if (var == NULL || strcmp("all", var) == 0) {
-    config_print_instr = val;
-    config_print_mem_access = val;
-    config_print_reg = val;
-    config_print_platform = val;
-    config_print_rvfi = val;
-  } else if (strcmp("instr", var) == 0) {
-    config_print_instr = val;
-  } else if (strcmp("reg", var) == 0) {
-    config_print_reg = val;
-  } else if (strcmp("abi", var) == 0) {
-    config_use_abi_names = val;
-  } else if (strcmp("mem", var) == 0) {
-    config_print_mem_access = val;
-  } else if (strcmp("rvfi", var) == 0) {
-    config_print_rvfi = val;
-  } else if (strcmp("platform", var) == 0) {
-    config_print_platform = val;
-  } else if (strcmp("step", var) == 0) {
-    config_print_step = val;
-  } else {
-    fprintf(stderr, "Unknown trace category: '%s' (should be %s)\n", var,
-            "instr|reg|mem|rvfi|platform|step|all");
-    exit(EXIT_FAILURE);
-  }
-}
+std::map<std::string, std::tuple<bool *, std::string>> printers {
+    {"instr",    {&config_print_instr, "Instruction execution"}     },
+    {"reg",      {&config_print_reg, "Register accesses"}           },
+    {"mem",      {&config_print_mem_access, "Memory accesses"}      },
+    {"rvfi",     {&config_print_rvfi, "RVFI tracing"}               },
+    {"platform",
+     {&config_print_platform, "Privilege changes, MMIO, interrupts"}},
+    {"step",     {&config_print_step, "Steps"}                      },
+    {"all",      {NULL, "All of the other tracers"}                 },
+};
+
+std::function<void(const std::string &)> enable_printer {
```

Comment:
Could these not be normal functions? This seems weird.

Reply:
The API does require a `std::function`
https://github.com/nadime15/sail-riscv/blob/d5c0fe197e973e8061c1dae8190675041f742c3a/dependencies/CLIUtils/CLI11.hpp#L7229

Reply:
Yeah but I think `std::function` can be implicitly constructed from a normal function pointer - try it!

Reply:
Yeah I guess this is slightly beyond C++'s template deduction rules. I pushed a commit that does it with an explicit `<std::string>` - hope that's ok!

Also I pushed a commit that allows `--trace` and `--no-trace` to be used more than once, which the old interface allowed.

Gotta say I'm not super impressed with CLI11. Definitely better than argparse but things like reflowing the footer... seems pretty amateur. Maybe I've just been spoiled by Rust's `clap` and Python's `Typer` which are both really really good.

Anyway LGTM with these changes.

Reply:
> Yeah I guess this is slightly beyond C++'s template deduction rules. I pushed a commit that does it with an explicit `<std::string>` - hope that's ok!
> 
> Also I pushed a commit that allows `--trace` and `--no-trace` to be used more than once, which the old interface allowed.

Great, thanks!